### DISCUSSION TITLE
P?

### DISCUSSION THREAD

**Discussion by ajmalton** [DivinumOfficium/divinum-officium#2618] [Developers corner]
Does any one know what the P in _Pofficium_ stands for? Pocket? PalmPilot? _Parvo_? or maybe just Phone?

There really is almost no difference at all between Pofficium and Officium.  At least some of the visual difference is browser rendering side effects. Almost all the rest is that Pofficium seems averse to popups.

The user can't choose different display options between the two, the only reason for having the difference as far as I can see.

I'm certainly to obsolete the file Pofficium.pl, and merge the few lines of difference into officium.pl, with a parameter like ?m=1 .

**Comment by gregordick** [comment]
I think that "Phone" is as good a guess as any!  +1 for merging it with `officium.pl`.

**Comment by fiapps** [comment]
Since `Pofficium.pl` is my baseline for integrating changes from the Divinum Officium Project, I would not be happy to see it disappear. If you're going to declare something obsolete, my vote would be for `officium.pl` with its archaic web interface.

**Comment by ajmalton** [comment]
I hear you. 

So are you actually _using_ the Pofficium interface? Somehow I thought you actually emit cleaned up HTML, scrape out the actual content, and wrap it. I didn't realize the Pofficium interface was important as such.

And surely both interfaces are archaic: obsolete html barely tolerated by modern browsers;  user-specified  screen geometry; server side navigation and calendar calculation; tiny unlabeled boxes; radio buttons that behave like hyperlinks; homebrew style-language; etc. etc.

But in fact the difference between the two officium entry points is very small, and the difference between the two UIs is half handled by conditionals in _webdia_ already.  So what I'm proposing is to use the same conditionals in officium.pl, in maybe four places, so that the remaining five or six hundred lines of code don't need to be duplicated and maintained twice.  No removal of functionality.  I suppose the result could be called Pofficium instead, if it matters.

**Reply by fiapps** [reply]
Upon further review,  withdraw my objection.

No, I'm not actually using the Pofficium interface, but it's the basis for my own interface, which is basically Pofficium minus forms and cookie handling. I chose Pofficium because there was less undesired stuff to delete from it. I have now spent some time looking through a diff of officium and Pofficium, and it looks like there's nothing I really care about that differs between the two programs. I gather it's `$expand = 'all';` that eliminates the popups.

There are several differences between the two programs that are just inconsistencies, but which will make it slightly harder to merge diffs from `officium.pl` instead of diffs from `Pofficium.pl`. For example, the former imports `do_io.pl`, does the other imports, and then sets the encoding to UTF-8. The latter sets the encoding to UTF-8, does the other imports, and then imports `do_io.pl`. So if a change is made to the list of imports, the diff won't quite match, and I'll have to stop to think about why it doesn't match and make the appropriate changes.

A couple of differences look like they could be bug fixes that didn't make it into Pofficium:

- `our $litaniaflag = 0;` (officium line 54, nowhere in Pofficium)
- `our $octavam = '';    #to avoid duplication of commemorations` (officium line 194, nowhere in Pofficium)

A weird difference was this: why does Pofficium (line 182) pass the date to `precedence()`, while officium (line 191) does not pass any argument? Basically because Pofficium has a different name for that parameter (`date1` instead of `date`). But it's strange for fetching the date from the parameters to be sometimes the responsibility of the top-level Perl file, and sometimes the responsibility of the core code.

**Reply by ajmalton** [reply]
Do you use a different cookie scheme? Or none at all? 
More important: do you rely on the setupsave scheme instead?

**Reply by fiapps** [reply]
I do not use cookies. My interface is API-like, in that the HTML returned depends solely upon the parameters passed in the HTTP request.

**Reply by ajmalton** [reply]
Excellent, very clear.  Of course the core should do exactly that, too.

**Comment by ajmalton** [comment]
There are vestiges of features, including more comprehensive _expand_, and partial bug fixes like _octavam_, and errors like _browserdate_, that are all reasons why I want to eliminate Pofficium.pl and port the tiny set of differences into officium.pl. Some of those features might eventually disappear or be improved away, but I want to be able do that in **one** place, not **two**. #lazyprogrammer #DRY

As for the $date1 argument to _precedence_, it might have sorta been right at some point in the past, but a glance at the logic shows that it makes not a scrap of difference currently, and the argumentless call is the better form (for now).

Pofficium pulls the date parameter because it wants to do trivial date calculations on the server. For no good reason that I can see, and officium more sensibly does them on the client. I'd change that, except that it would change the UI at the HTTP and HTML levels, which I am at the moment trying to avoid.
